### PR TITLE
feat(middleware): add middleware package

### PR DESCRIPTION
**Pull request by RealShadowNova**
This PR adds the `@joshdb/middleware` package to add modularity and simplicity in the middleware creation process.

# Changes

- Add `packages/middleware/` directory
- Update root `README.md`
- Update the `test` script to build before running tests to ensure workspace linking

### CODE REVIEW

**Review by dan-online — DISMISSED** [review]
LGTM although in future provider and serialize changes should probably have been in their own pr

**Review by dan-online — APPROVED** [review]
LGTM but checks failed

File: packages/middleware/README.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+## Description
+
+A core Josh utility to create middleware.
+
+## Features
+
+- Written in TypeScript
+- Offers CommonJS, ESM, and UMD bundles
+- Fully tested
```
Comment (dan-online) [inline packages/middleware/README.md:22]: Can we discuss this features section, its repetitive in every repo and doesn't contain every cool feature

File: packages/middleware/README.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+## Installation
+
+You can use the following command to install this package, or replace `npm install` with your package manager of choice.
+
+```sh
+npm install @joshdb/middleware
+```
```
Comment (dan-online) [inline packages/middleware/README.md:30]: Add yarn to keep consistency across readmes

**Review by dan-online — CHANGES_REQUESTED** [review]
See comments

File: packages/middleware/README.md (hunk 2 above)
Comment (RealShadowNova) [reply packages/middleware/README.md:30]: It is consistent?

File: packages/middleware/README.md (hunk 1 above)
Comment (RealShadowNova) [reply packages/middleware/README.md:22]: Yeah, basically just placeholders, we could probably do this better later on when we start working on website

File: packages/middleware/README.md (hunk 2 above)
Comment (dan-online) [reply packages/middleware/README.md:30]: I see it in @josh-development/core

Comment (RealShadowNova) [reply packages/middleware/README.md:30]: That's to be done

**Pull request merged**